<a href="https://colab.research.google.com/github/EEdwardsA/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment_Elizabeth.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [ ]:
### YOUR WORK HERE

import pandas as pd
import numpy as np

In [ ]:
columns = ['symboling','normalized-losses','make','fuel-type','aspiration',
           'num-of-doors','body-style','drive-wheels','engine-location',
           'wheel-base','length','width','height','curb-weight','engine-type',
           'num-of-cylinders','engine-size','fuel-system','bore','stroke',
           'compression','horsepower','peak-rpm','city-mpg','hwy-mpg','price']

df = pd.read_csv('imports-85.data', names=columns)

In [ ]:
df.head()

symboling normalized-losses         make  ... city-mpg hwy-mpg  price
0          3                 ?  alfa-romero  ...       21      27  13495
1          3                 ?  alfa-romero  ...       21      27  16500
2          1                 ?  alfa-romero  ...       19      26  16500
3          2               164         audi  ...       24      30  13950
4          2               164         audi  ...       18      22  17450

[5 rows x 26 columns]

In [ ]:
df = df.replace({'?':np.NaN})

df.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression           0
horsepower            2
peak-rpm              2
city-mpg              0
hwy-mpg               0
price                 4
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [ ]:
### YOUR WORK HERE

# Check dtypes to make sure highway-mpg is numerical

df['hwy-mpg'].dtypes

dtype('int64')

In [ ]:
sample = df.sample(10,random_state=30)

sample 

symboling normalized-losses           make  ... city-mpg hwy-mpg  price
121          1               154       plymouth  ...       31      38   6692
176         -1                65         toyota  ...       27      32  10898
24           1               148          dodge  ...       31      38   6229
88          -1               137     mitsubishi  ...       23      30   9279
72           3               142  mercedes-benz  ...       16      18  35056
103          0               108         nissan  ...       19      25  13499
26           1               148          dodge  ...       31      38   7609
20           0                81      chevrolet  ...       38      43   6575
155          0                91         toyota  ...       27      32   8778
185          2                94     volkswagen  ...       27      34   8195

[10 rows x 26 columns]

In [ ]:
df['hwy-mpg'].mean()

30.75121951219512

In [ ]:
from scipy import stats

In [ ]:
stats.ttest_1samp(sample['hwy-mpg'], 30.75121951219512)

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

P = .95

The null hypothesis is that the sample is the same as the population.

Alternative hypothesis: the miles per gallon on the highway for the sample is different from that of the larger population.

Based on a t-statistic of 0.8912, and a p-value of 0.39, we fail to reject the null hypothesis that the sample cars have a different highway MPG than all cars overall.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [ ]:
### YOUR WORK HERE

# Null hypothesis: salesman is telling the truth, cars I test have average mpg of 35.

stats.ttest_1samp(sample['hwy-mpg'],35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

P value is not = or < .05. I fail to reject the null hypothesis. I do not reject the salesman's claim.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [ ]:
### YOUR WORK HERE

larger_sample = df.sample(100,random_state=30)

stats.ttest_1samp(larger_sample['hwy-mpg'],35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

In the second case, with a larger sample, I reject the null hypothesis and suggest the alternative hypothesis, that the salesman's cars *do not* have an average highway MPG of 35. 

The two t-tests have different results because in the second t-test, the larger sample size gives us more information, which gives us more certainty and more ability to draw conclusions about probability.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
### YOUR WORK HERE

column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('house-votes-84.data', names= column_headers)

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [ ]:
df = df.replace({'y':1,'n':0,'?':np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [ ]:
rep = df[df['party']=='republican']
dem = df[df['party']=='democrat']

rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [ ]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [ ]:
### TEST 1
# Null hypothesis: Democrats and Republicans voted the same way on the budget bill.
# Alternative hypothesis: Dems and Repubs voted differently on the budget bill.

stats.ttest_ind(dem['budget'],rep['budget'],nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

In [ ]:
#Conclusion: Based on a t-statistic of 23.2 and a p-value too small to type out,
# we reject the null hypothesis and suggest the alternative hypothesis, that 
# Republicans and Democrats voted differently on the 1984 budget bill.
# Because the t-statistic has a positive sign, Democrats supported the budget bill
# more than Republicans.

### TEST 2
# Null hypothesis: Democrats and Republicans voted the same way on the 
# 'physician fee freeze' bill.
# Alternative hypothesis: Dems and Repubs voted differently on the 'physician 
# fee freeze' bill.

stats.ttest_ind(dem['physician-fee-freeze'],rep['physician-fee-freeze'],
                nan_policy='omit')


Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177)

In [ ]:
# Conclusion: Based on a t-statistic of -49.4 and a p-value of 1.99 to the negative
# 177th power, we reject the null hypothesis and suggest the alternative hypothesis,
# that Republicans and Democrats voted differently on a bill regarding freezing
# physician fees. Because the t-statistic has a negative sign, we can tell Republicans
# supported the bill more than Democrats.

###TEST 3
# Null hypothesis: Democrats and Republicans voted the same way on the water
# project bill.
# Alternative hypothesis: Dems and Repubs voted differently on the water project
# bill.

# p-value < .10 -> reject

# P-value >= .10 -> fail to reject

stats.ttest_ind(dem['water-project'],rep['water-project'],
                nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

In [ ]:
# Conclusion: Based on a t-statistic of -0.089 and a p-value of 0.929, I fail to
# reject the null hypothesis that Democrats and Republicans voted the same on
#the water project bill.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE